### PR TITLE
BUG: #57775 Fix groupby apply in case func returns None for all groups

### LINKED ISSUE #57775: BUG: groupby.apply when func always returns None returns empty dataframe
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

print(pd.DataFrame(['a']).groupby(0).apply(lambda x: None))
```

### Issue Description

Returns an empty dataframe with no columns or index

### Expected Behavior

Should return something like 

```python
0
a   None
dtype: object
```

### Installed Versions

<details>

```

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.9.18.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.3.0
Version               : Darwin Kernel Version 23.3.0: Wed Dec 20 21:31:00 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6020
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.1
numpy                 : 1.26.3
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : 8.18.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.4
qtpy                  : None
pyqt5                 : None

```

</details>

### PR DESCRIPTION
**Pull request by dontgoto**
- [x] closes #57775
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.2.rst` file if fixing a bug or adding a new feature.

In case the func passed to DataFrameGroupBy.apply returns None for all groups (filtering out all groups), the original dataframe's columns and dtypes are now part of the empty result dataframe.

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
seems reasonable to me. @rhshadrach

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1572,6 +1572,13 @@ def apply(self, func, *args, include_groups: bool = True, **kwargs) -> NDFrameT:
         behavior or errors and are not supported. See :ref:`gotchas.udf-mutation`
         for more details.
 
+        Groups for which ``func`` returns ``None`` will be filtered from the result.
+
+        .. versionchanged:: 2.2.2
+
+            In case all groups are filtered from the result, an empty DataFrame
+            with the columns and dtypes of the original dataframe will be returned.
+
```
Comment (mroeschke) [inline pandas/core/groupby/groupby.py:1581]:
```suggestion
```

(the example below is sufficient)

File: doc/source/whatsnew/v2.2.2.rst [hunk 2]
```diff
@@ -22,7 +22,6 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 -
-
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.2.rst:25]: Could you undo this white space change?

File: doc/source/whatsnew/v2.2.2.rst [hunk 3]
```diff
@@ -21,7 +21,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- :meth:`DataFrameGroupBy.apply` was returning a completely empty DataFrame when all return values of ``func`` were ``None`` instead of returning an empty DataFrame with the original columns and dtypes. (:issue:`57775`)
```
[the remark this answers is not in the document]
Comment (dontgoto) [reply doc/source/whatsnew/v2.2.2.rst]: Done

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (dontgoto) [reply pandas/core/groupby/groupby.py]: Removed the description, only left the example

File: doc/source/whatsnew/v2.2.2.rst (hunk 2 above)
Comment (dontgoto) [reply doc/source/whatsnew/v2.2.2.rst]: Done. Thanks, I missed that one.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @dontgoto

**Pull request merged**